### ISSUE TITLE
use memory prometheus 

### ISSUE THREAD

**Issue by siddontang** [pingcap/tikv#793]
We have already used prometheus in pd, so we decide to use this in tikv. 
Maybe we can use https://github.com/posix4e/rust-metrics or other better solutions. 

First the memory metric is ok, and like rocksdb, we can supply a interval configuration, like 1m, for every 1m, we will output the metric to log. 

Hi @overvenus could you help us to support this?

**Comment by overvenus** [comment]
@siddontang Sure

**Comment by overvenus** [comment]
rust-metrics for prometheus is not ready yet, we should wait untill it becomes beta. I will send some PRs to accelerate the process.

**Comment by siddontang** [comment]
where not ready?
We don't need to send metrics to prometheus now. only memory is ok, can this be supported yet?

**Comment by overvenus** [comment]
OK, I will output metrics to log for now.

**Comment by ngaut** [comment]
Do yo need any help ? @overvenus

**Comment by overvenus** [comment]
@ngaut To be frank, I do.

rust-metric requires that we should register metrics first before recording them. Once reporters(A struct which sends metrics to prometheus or log) are started, we can't make any new metrics.

```
counter1
counter2
register("c1", counter1)

start reporter

counter1.inc() // OK
counter2.inc() // NO, the count will not be sent
register("c2", counter2) // NO, reporter has started
```

So we need register all metrics early in the runtime of the program. Also I do not think macro fits in rust-metric as it records metrics by variable not string

```
counter1.inc()

// NOT
counter_inc!('foo')
```  

BTW, rust-metric master branch have changed many place compares to v0.1.1 in crate.io, but the master branch can not be compiled unless we generate `prome_proto.rs`(prometheus protobuf file) manually. I will send a PR to fix this problem later.

Sorry about the wall of text :sweat_smile:

**Comment by siddontang** [comment]
Hi @overvenus 

You can see the prometheus go client and will find that it needs registering first too. So I think it is ok here.

**Comment by overvenus** [comment]
Closed via #1017
